### ISSUE TITLE
systemd-networkd DHCPServer should select static IP addresses for clients from /etc/ethers

### ISSUE THREAD

**Issue by dkg** [systemd/systemd#18291]
I'm using the DHCPServer component of systemd-networkd.  Some clients on the network should be assigned static IP addresses. Typically (with other DHCP servers) this is done by associating MAC addresses with dedicated IP addresses so that the DHCP server can emit a stable IP address for specific machines.    It is not clear that there is any way to make such an assignment with the DHCPServer component of systemd-networkd.

I think the simplest thing would be to have it source MAC address → IP address bindings from `/etc/ethers` (see `ethers(5)`), using the following approach:

 - if the DHCP client's MAC address is in `/etc/ethers`, and
 - it maps to an IP address X that is within the subnet configured for the server address (note that this could be an address outside of the pool itself, if PoolOffset or PoolSize are set), and
 - that address is not currently assigned to a different device, then
 - offer a lease for IP address X to the DHCP client.

If any of the above conditions does not hold, allocate an address from the preexisting pool.

For additional robustness, `systemd-networkd` could pre-emptively scan all of `/etc/ethers` to determine which IP addresses within the pool itself may be associated with a static device;  With that list assembled, the DHCPServer component could avoid offering leases on those IP addresses unless the rest of the pool is already allocated.

While a baseline implementation of this functionality could just scan `/etc/ethers` once at startup, this would mean that the system administrator would need to restart systemd-networkd when `/etc/ethers` changes.  This sounds fine to me, given the need to restart `systemd-networkd` for certain other configuration changes as well.

But a final bit of nice engineering could be to consider how/when to re-read `/etc/ethers` during operation.  If this is desired, two options seem possible (i have no preference between them):

 - the DHCPServer component could stat `/etc/ethers` at startup and cache the result; upon receiving a new DHCP request, it could stat it again to see whether the file has changed; if it has, rescan it and update the internal state accordingly before responding to the DHCP request.
 - the DHCPServer component could register an inotify handler listener to watch for changes in `/etc/ethers`.  If a change appears, it could rescan the file and update the internal state accordingly.

**Comment by bluca** [comment]
There's a PR under review implementing static leases: https://github.com/systemd/systemd/pull/15556

**Comment by poettering** [comment]
hmm, /etc/ethers, really? is *anything* actually using that? To me that smells like left-over cruft from the 1970, that we probably shouldn't breathe new life into, but just let die. I mean /etc/services is someting we don't support either...

Thing is, /etc/ethers is way to simple for defining static DHCP lease info, since you probably want to map more than just a MAC address, but also all kinds of other DHCP lease params...

**Comment by dkg** [comment]
It's certainly leftover from the 1970s, but that doesn't mean it's the worst thing to use.  the proposal over on #15556 offers a different configuration mechanism for declaring static leases (with quite a bit more code than just parsing `/etc/ethers` i suspect) but still doesn't offer any additional mapping possibilities beyond a basic MAC→IP table.  I'd be happy with either configuration mechanism, though!

**Comment by poettering** [comment]
> It's certainly leftover from the 1970s, but that doesn't mean it's the worst thing to use. the proposal over on #15556 offers a different configuration mechanism for declaring static leases (with quite a bit more code than just parsing `/etc/ethers` i suspect) but still doesn't offer any additional mapping possibilities beyond a basic MAC→IP table. I'd be happy with either configuration mechanism, though!

True, the mechanism in #15556 doesn't allow configuring more than just the IP address, but the design of the concept allows natural extension to allow that. /etc/ethers doesn't really, it's defined the way it is, and not extensible by us.

BTW, apparently other DHCP server implementions support checking that, so my question regarding who uses that stuff is answered: dnsmasq supports it and the ISC server too, afaics.

**Comment by poettering** [comment]
glibc has functions to access this stuff even: `ether_hostton()`. and even supports NSS for it! crazy shit. ;-)

**Comment by yuwata** [comment]
I like this proposal. We already support /etc/hosts in resolved.

**Comment by bluca** [comment]
> glibc has functions to access this stuff even: `ether_hostton()`. and even supports NSS for it! crazy shit. ;-)

I've done a few years stint in the networking OSes world, and never came across anything _really_ using this as far as I can remember... maybe if folks really want it, it could be handled somewhat like resolv.conf is handled by resolved?
But it should be something to add to the TODO, rather than a blocker for this PR, IMHO.

**Comment by poettering** [comment]
Seems to be quite a bit of software that actually reads that stuff:

https://codesearch.debian.net/search?q=%2Fetc%2Fethers&literal=1

> I like this proposal. We already support /etc/hosts in resolved.

That is true, but there are difference: `/etc/hosts` is *really* well known and established. 

But dunno, up to you if you want to use that.

(We shouldn't use the glibc API for this though, given it uses NSS it might be backed by LDAP/SSSD or so, i.e. be async and slow, and hence not easy to call. Also, doesn't allow indexing/caching like we do for /etc/hosts)

**Comment by dkg** [comment]
Thinking on this further, one more downside with `/etc/ethers` is that it's technically defined such that the second entry on each line can be a hostname, not just an IP address (`ethers(5)` says, confusingly: "the *IP-number* may be a hostname which can be resolved by DNS or a dot separated number").

If an admin uses a DNS name there, then resolving that ends up introducing yet another layer of delay, possible failure, and staleness to any information read from it.

**Comment by poettering** [comment]
hmm, the fact that the rhs can be a hostname makes it kinda unsuitable for the purpose, or only in a very skewed fasion...

doing network-based name resolution on this stuff sucks hard, we need quick, synchronous access to this database or not do it at al...
